--- src_radiation.f90	2014-11-28 12:18:36.000000000 +0100
+++ src_radiation.f90	2019-12-04 12:52:07.114213000 +0100
@@ -445,6 +445,11 @@
 
 !------------------------------------------------------------------------------
 
+#ifdef COUP_OAS_COS
+USE data_fields, ONLY :       &
+     trad_clm         ! CLM radiative temperature as bottom BC
+#endif
+
 USE data_runcontrol , ONLY :   &
 
 ! 1. start and end of the forecast
@@ -454,6 +459,7 @@
     nold  ,       & ! corresponds to ntstep - 1
     nnow  ,       & ! corresponds to ntstep 
     nnew  ,       & ! corresponds to ntstep + 1
+    hstart,       & 
 
 ! 3. controlling the physics
 ! --------------------------
@@ -1528,6 +1534,9 @@
   zsmu0  (ie,je )        , & ! Cosine of zenith angle 
   zalso  (ie,je )        , & ! Solar surface albedo
   zalth  (ie,je )        , & ! Thermal surface albedo 
+#ifdef COUP_OAS_COS
+  zpalp  (ie,je )        , & ! Direct albedo
+#endif
   zskyview  (ie,je)      , & ! used as argument for SR fesft
 
   ! other values for intermediate storage
@@ -1608,6 +1617,9 @@
   zapre_rn  (:  )      , & ! Surface pressure
   zsmu0_rn  (:  )      , & ! Cosine of zenith angle
   zalth_rn  (:  )      , & ! Thermal surface albedo
+#ifdef COUP_OAS_COS
+  zpalp_rn  (:  )      , & ! Solar surface parallel albedo
+#endif
   zalso_rn  (:  )          ! Solar surface albedo
 
   REAL (KIND=wp)       , ALLOCATABLE :: &
@@ -2333,6 +2345,8 @@
       ENDIF
     ENDDO
 
+#ifndef COUP_OAS_COS
+
     IF (lradave) THEN
       DO j = jstartrad, jendparrad      ! jstartpar, jendpar
         DO  i = istartrad, iendparrad
@@ -2341,6 +2355,27 @@
       ENDDO
     ENDIF
 
+#else
+!------------------------------------------------------------------------------
+! provide Community Land Model surface albedo instead of albedo above
+! Only for land points (llandmask)
+!------------------------------------------------------------------------------
+  DO j = jstartrad, jendparrad      ! jstartpar, jendpar
+    DO  i = istartrad, iendparrad
+      IF (llandmask(i,j)) THEN
+         zpalp (i,j) = alb_rad(i,j,1)
+         zalso (i,j) = alb_rad(i,j,2)
+      ELSE
+!CPS need this for parallel radiation
+        zpalp  (i,j) = (1.0_wp +                                           &
+         0.5_wp * ((1.0_wp/zsmu0(i,j)) * (1.0_wp/zalso(i,j) - 1.0_wp))) &
+      / (1.0_wp + ((1.0_wp/zsmu0(i,j)) * (1.0_wp/zalso(i,j) - 1.0_wp)))**2
+!CPS need this for parallel radiation
+      ENDIF !   llandmask
+    ENDDO
+  ENDDO
+#endif
+
     !------------------------------------------------------------------------------
     ! Section 3:  Set cloudiness and humidity on input for fesft; 
     !             Store cloud cover on corresponding global arrays
@@ -2953,6 +2988,9 @@
     ALLOCATE ( zsmu0_rn  (idim_rad    ) , STAT=izstata )
     ALLOCATE ( zalth_rn  (idim_rad    ) , STAT=izstata )
     ALLOCATE ( zalso_rn  (idim_rad    ) , STAT=izstata )
+#ifdef COUP_OAS_COS
+    ALLOCATE ( zpalp_rn  (idim_rad    ) , STAT=izstata )
+#endif
     ALLOCATE ( zskyv_rn  (idim_rad    ) , STAT=izstata )
     ALLOCATE ( zscor_rn  (idim_rad    ) , STAT=izstata )
 
@@ -2989,6 +3027,9 @@
     zaeq4_rn(:,:)   = 0.0_wp
     zaeq5_rn(:,:)   = 0.0_wp
     zalso_rn(:)     = 0.0_wp
+#ifdef COUP_OAS_COS
+    zpalp_rn(:)     = 0.0_wp
+#endif
     zalth_rn(:)     = 0.0_wp
     zapre_rn(:)     = 0.0_wp
     zsmu0_rn(:)     = 0.0_wp
@@ -3077,6 +3118,18 @@
           zti_rn    (i,ke1) = zti    (i,js,ke1)
 
           zalso_rn  (i)     = zalso    (i,js)
+#ifdef COUP_OAS_COS
+
+           IF (( ntstep * dt ) - (  hstart * 3600.0 )>0) THEN        !At start, trad_clm = 0.0
+             zti_rn    (i,ke1) = trad_clm (i,js)   !LongWave Closure CPS
+           ELSEIF (hstart > 0 .and. ( ntstep * dt ) - (  hstart * 3600.0 ) == 0) THEN
+             zti_rn    (i,ke1) = zti    (i,js,ke1)
+           ELSE
+             zti_rn    (i,ke1) = zti    (i,js,ke1)
+           ENDIF
+
+          zpalp_rn  (i)     = zpalp    (i,js)
+#endif 
           zalth_rn  (i)     = zalth    (i,js)
           zapre_rn  (i)     = zapre    (i,js)
           zsmu0_rn  (i)     = zsmu0    (i,js)
@@ -3086,6 +3139,20 @@
       ENDIF
 
       IF (lrady) THEN
+#ifdef COUP_OAS_COS 
+        CALL fesft                                                          &
+          (zti_rn,    zdpr_rn,     zclc_rn,     zwv_rn,    zsw_rn,          &
+           zclwc_rn,  zciwc_rn,    zduco2f_rn,  zduo3f_rn,                  &
+           zaeq1_rn,  zaeq2_rn,    zaeq3_rn,    zaeq4_rn,  zaeq5_rn,        &
+           zapre_rn,  zsmu0_rn,    zalso_rn,    zalth_rn,  zskyv_rn,        &
+           zscor_rn,  zstb,        zsct,        zpalp_rn,                   &
+           ki1sd,     ki1ed,                               ki3sd,    ki3ed, &
+           ki1sc,     ki1ec,                               ki3sc,    ki3ec, &
+           lsolar(js),lcrf,        lradtopo,    izdebug,   js,              &
+           zflt_rn,   zfls_rn,     zflt_s_rn,   zfls_s_rn, zflsdir_rn,      &
+           zfltd_rn,  zfltu_rn,    zflsd_rn,    zflsu_rn,  zflsp_rn,        &
+           zflpar_rn, zflsu_par_rn, zflsd_par_rn, zflsp_par_rn)
+#else
         CALL fesft                                                          &
           (zti_rn,    zdpr_rn,     zclc_rn,     zwv_rn,    zsw_rn,          &
            zclwc_rn,  zciwc_rn,    zduco2f_rn,  zduo3f_rn,                  &
@@ -3098,6 +3165,7 @@
            zflt_rn,   zfls_rn,     zflt_s_rn,   zfls_s_rn, zflsdir_rn,      &
            zfltd_rn,  zfltu_rn,    zflsd_rn,    zflsu_rn,  zflsp_rn,        &
            zflpar_rn, zflsu_par_rn, zflsd_par_rn, zflsp_par_rn)
+#endif
       ENDIF
 
       IF (lradave) THEN
@@ -3174,6 +3242,9 @@
           zaeq4_rn(:,:)   = 0.0_wp
           zaeq5_rn(:,:)   = 0.0_wp
           zalso_rn(:)     = 0.0_wp
+#ifdef COUP_OAS_COS
+          zpalp_rn(:)     = 0.0_wp
+#endif
           zalth_rn(:)     = 0.0_wp
           zapre_rn(:)     = 0.0_wp
           zsmu0_rn(:)     = 0.0_wp
@@ -3248,7 +3319,12 @@
           zzflsd_par(i,js) = zflsd_par(i)
           zzflsu_par(i,js) = zflsu_par(i)
 
-          alb_rad (i,js) = zalso (i,js)
+#ifdef COUP_OAS_COS
+          alb_rad (i,js,1) = zpalp (i,js)
+          alb_rad (i,js,2) = zalso (i,js)
+#else
+          alb_rad (i,js)   = zalso (i,js)
+#endif
           
         ENDDO
 
@@ -3295,6 +3371,9 @@
     DEALLOCATE ( zsmu0_rn     , STAT=izstatd )
     DEALLOCATE ( zalth_rn     , STAT=izstatd )
     DEALLOCATE ( zalso_rn     , STAT=izstatd )
+#ifdef COUP_OAS_COS
+    DEALLOCATE ( zpalp_rn     , STAT=izstatd )
+#endif
     DEALLOCATE ( zflt_rn      , STAT=izstatd )
     DEALLOCATE ( zfls_rn      , STAT=izstatd )
     DEALLOCATE ( zflpar_rn    , STAT=izstatd )
@@ -3722,7 +3801,11 @@
       DO i = istartpar,iendpar
         ! such it has been tested in testsuite 3.4
         ! but keep it for the moment to reproduce results with Version 3.22
+#ifdef COUP_OAS_COS
+        zalbfak = (1.0_wp-alb_rad(i,js,2))
+#else
         zalbfak = (1.0_wp-alb_rad(i,js))
+#endif
         sobs (i,js) = sobs(i,js) * zalbfak
         pabs (i,js) = pabs(i,js) * zalbfak
         ! And this seems to be the better version
@@ -3987,6 +4070,9 @@
         i = istartrad + (n-1) * nradcoarse
         zti_rn(n,ke+1) = zti_rn(n,ke+1) + zti  (i,js-1,ke+1) + zti  (i+1,js-1,ke+1) + zti  (i,js,ke+1) + zti  (i+1,js,ke+1)
         zalso_rn(n)    = zalso_rn(n)    + zalso(i,js-1)      + zalso(i+1,js-1)      + zalso(i,js)      + zalso(i+1,js)
+#ifdef COUP_OAS_COS
+        zpalp_rn(n)    = zpalp_rn(n)    + zpalp(i,js-1)      + zpalp(i+1,js-1)      + zpalp(i,js)      + zpalp(i+1,js)
+#endif
         zalth_rn(n)    = zalth_rn(n)    + zalth(i,js-1)      + zalth(i+1,js-1)      + zalth(i,js)      + zalth(i+1,js)
         zapre_rn(n)    = zapre_rn(n)    + p0hl (i,js-1,ke+1) + p0hl (i+1,js-1,ke+1) + p0hl (i,js,ke+1) + p0hl (i+1,js,ke+1)
         zsmu0_rn(n)    = zsmu0_rn(n)    + zsmu0(i,js-1)      + zsmu0(i+1,js-1)      + zsmu0(i,js)      + zsmu0(i+1,js)
@@ -3996,6 +4082,9 @@
         i = istartrad + (n-1) * nradcoarse
         zti_rn  (n,ke+1) = zti_rn  (n,ke+1) + zti  (i,js,ke+1) + zti  (i+1,js,ke+1)
         zalso_rn(n)      = zalso_rn(n)      + zalso(i,js)      + zalso(i+1,js)
+#ifdef COUP_OAS_COS
+        zpalp_rn(n)      = zpalp_rn(n)      + zpalp(i,js)      + zpalp(i+1,js)
+#endif
         zalth_rn(n)      = zalth_rn(n)      + zalth(i,js)      + zalth(i+1,js)
         zapre_rn(n)      = zapre_rn(n)      + p0hl (i,js,ke+1) + p0hl (i+1,js,ke+1)
         zsmu0_rn(n)      = zsmu0_rn(n)      + zsmu0(i,js)      + zsmu0(i+1,js)
@@ -4004,6 +4093,9 @@
     DO n = 1, izz
       zti_rn  (n,ke+1) = zti_rn  (n,ke+1) * zfactor
       zalso_rn(n)      = zalso_rn(n)      * zfactor
+#ifdef COUP_OAS_COS
+      zpalp_rn(n)      = zpalp_rn(n)      * zfactor
+#endif
       zalth_rn(n)      = zalth_rn(n)      * zfactor
       zapre_rn(n)      = zapre_rn(n)      * zfactor
       zsmu0_rn(n)      = zsmu0_rn(n)      * zfactor
@@ -4016,18 +4108,27 @@
       IF (nradcoarse_y == 2) THEN
         zti_rn  (izz,ke+1) = zti_rn  (izz,ke+1) + zti  (i,js-1,ke+1) + zti  (i,js,ke+1)
         zalso_rn(izz)      = zalso_rn(izz)      + zalso(i,js-1)      + zalso(i,js)
+#ifdef COUP_OAS_COS
+        zpalp_rn(izz)      = zpalp_rn(izz)      + zpalp(i,js-1)      + zpalp(i,js)
+#endif
         zalth_rn(izz)      = zalth_rn(izz)      + zalth(i,js-1)      + zalth(i,js)
         zapre_rn(izz)      = zapre_rn(izz)      + p0hl (i,js-1,ke+1) + p0hl (i,js,ke+1)
         zsmu0_rn(izz)      = zsmu0_rn(izz)      + zsmu0(i,js-1)      + zsmu0(i,js)
       ELSE  ! nradcoarse_y == 1
         zti_rn  (izz,ke+1) = zti_rn  (izz,ke+1) + zti  (i,js,ke+1)
         zalso_rn(izz)      = zalso_rn(izz)      + zalso(i,js)
+#ifdef COUP_OAS_COS
+        zpalp_rn(izz)      = zpalp_rn(izz)      + zpalp(i,js)
+#endif
         zalth_rn(izz)      = zalth_rn(izz)      + zalth(i,js)
         zapre_rn(izz)      = zapre_rn(izz)      + p0hl (i,js,ke+1)
         zsmu0_rn(izz)      = zsmu0_rn(izz)      + zsmu0(i,js)
       ENDIF !
       zti_rn  (izz,ke+1) = zti_rn  (izz,ke+1) * zfactor_b
       zalso_rn(izz)      = zalso_rn(izz)      * zfactor_b
+#ifdef COUP_OAS_COS
+      zpalp_rn(izz)      = zpalp_rn(izz)      * zfactor_b
+#endif
       zalth_rn(izz)      = zalth_rn(izz)      * zfactor_b
       zapre_rn(izz)      = zapre_rn(izz)      * zfactor_b
       zsmu0_rn(izz)      = zsmu0_rn(izz)      * zfactor_b
@@ -4117,6 +4218,9 @@
       zflt_s   (i+ii)     = zflt_s_rn   (izz)
       zfls_s   (i+ii)     = zfls_s_rn   (izz)
       zalso    (i+ii,js)  = zalso_rn    (izz)
+#ifdef COUP_OAS_COS
+      zpalp    (i+ii,js)  = zpalp_rn    (izz)
+#endif
       tg_rn    (i+ii)     = zti_rn      (izz,ke1)
       zfltu    (i+ii)     = zfltu_rn    (izz)
       zfltd    (i+ii)     = zfltd_rn    (izz)
@@ -4139,6 +4243,9 @@
         zflt_s   (i+ii)     = zflt_s_rn   (izz)
         zfls_s   (i+ii)     = zfls_s_rn   (izz)
         zalso    (i+ii,js)  = zalso_rn    (izz)
+#ifdef COUP_OAS_COS
+        zpalp    (i+ii,js)  = zpalp_rn    (izz)
+#endif
         tg_rn    (i+ii)     = zti_rn      (izz,ke1)
         zfltu    (i+ii)     = zfltu_rn    (izz)
         zfltd    (i+ii)     = zfltd_rn    (izz)
@@ -4169,6 +4276,21 @@
 !==============================================================================
 !+ Module procedure in "Radiation" 
 !------------------------------------------------------------------------------
+#ifdef COUP_OAS_COS
+!CPS we receive albedo for direct radiation also (palp)
+SUBROUTINE fesft_sp(                                                     &
+       pti   , pdp  , pclc  , pwv  , psw  , pqlwc, pqiwc, pduco2, pduo3, &
+       paeq1 , paeq2, paeq3 , paeq4, paeq5,                              &
+       papre , psmu0, palso , palth, pskyview, pfcor,                    &
+       psig  , psct , palp  ,                                            &
+       ki1sd , ki1ed,                ki3sd, ki3ed,                       &
+       ki1sc , ki1ec,                ki3sc, ki3ec,                       &
+       lsolar, lcrf , lradtopo, idebug, jindex,                          &
+       pflt  , pfls , pflt_s, pfls_s, pflsdir,                           &
+       pfltd , pfltu, pflsd , pflsu, pflsp,                              &
+       pflpar, pflsu_par, pflsd_par, pflsp_par)
+
+#else
 
 SUBROUTINE fesft_sp(                                                     &
        pti   , pdp  , pclc  , pwv  , psw  , pqlwc, pqiwc, pduco2, pduo3, &
@@ -4182,7 +4304,7 @@
        pfltd , pfltu, pflsd , pflsu, pflsp,                              &
        pflpar, pflsu_par, pflsd_par, pflsp_par)
 
-
+#endif
 !------------------------------------------------------------------------------
 !
 ! Description:
@@ -4250,7 +4372,9 @@
     psmu0  (ki1sd:ki1ed)                              , & ! (1)
     palth  (ki1sd:ki1ed)                              , & ! (1)
     palso  (ki1sd:ki1ed)                              , & ! (1)
-  
+#ifdef COUP_OAS_COS
+    palp   (ki1sd:ki1ed)                              , & ! (1)
+#endif 
     ! External data and radiation corrections
     pskyview(ki1sd:ki1ed)                             , & !
     pfcor   (ki1sd:ki1ed)            
@@ -4322,7 +4446,9 @@
     psmu0_dp  (ki1sd:ki1ed)                              , & ! (1)
     palth_dp  (ki1sd:ki1ed)                              , & ! (1)
     palso_dp  (ki1sd:ki1ed)                              , & ! (1)
-  
+#ifdef COUP_OAS_COS
+    palp_dp   (ki1sd:ki1ed)                              , & ! (1)
+#endif  
     ! External data and radiation corrections
     pskyview_dp(ki1sd:ki1ed)                             , & !
     pfcor_dp   (ki1sd:ki1ed)
@@ -4389,7 +4515,9 @@
   psmu0_dp   (:)    = 0.0_dp
   palth_dp   (:)    = 0.0_dp
   palso_dp   (:)    = 0.0_dp
-
+#ifdef COUP_OAS_COS
+  palp_dp    (:)    = 0.0_dp
+#endif 
   pskyview_dp(:)    = 0.0_dp
   pfcor_dp   (:)    = 0.0_dp
 
@@ -4427,20 +4555,23 @@
   psmu0_dp   (ki1sc:ki1ec)    = REAL(psmu0   (ki1sc:ki1ec),   dp)
   palth_dp   (ki1sc:ki1ec)    = REAL(palth   (ki1sc:ki1ec),   dp)
   palso_dp   (ki1sc:ki1ec)    = REAL(palso   (ki1sc:ki1ec),   dp)
-
+#ifdef COUP_OAS_COS
+  palp_dp    (ki1sc:ki1ec)    = REAL(palp    (ki1sc:ki1ec),   dp)
+#endif 
   pskyview_dp(ki1sc:ki1ec)    = REAL(pskyview(ki1sc:ki1ec),   dp)
   pfcor_dp   (ki1sc:ki1ec)    = REAL(pfcor   (ki1sc:ki1ec),   dp)
 
 ! Variable with intent inout (has to be transformed back at the end
   papre_dp   (ki1sc:ki1ec)    = REAL(papre   (ki1sc:ki1ec),   dp)
 
+#ifdef COUP_OAS_COS
 ! Now call double precision fesft
   CALL fesft_dp                                                               &
       (pti_dp,      pdp_dp,      pclc_dp,     pwv_dp,      psw_dp,            &
        pqlwc_dp,    pqiwc_dp,    pduco2_dp,   pduo3_dp,                       &
        paeq1_dp,    paeq2_dp,    paeq3_dp,    paeq4_dp,    paeq5_dp,          &
        papre_dp,    psmu0_dp,    palso_dp,    palth_dp,    pskyview_dp,       &
-       pfcor_dp,    psig_dp,     psct_dp,                                     &
+       pfcor_dp,    psig_dp,     psct_dp,     palp_dp,                        &
        ki1sd,       ki1ed,                                 ki3sd,      ki3ed, &
        ki1sc,       ki1ec,                                 ki3sc,      ki3ec, &
        lsolar,      lcrf,        lradtopo,    idebug,      jindex,            &
@@ -4448,6 +4579,21 @@
        pfltd_dp,    pfltu_dp,    pflsd_dp,    pflsu_dp,    pflsp_dp,          &
        pflpar_dp,   pflsu_par_dp, pflsd_par_dp, pflsp_par_dp)
 
+#else
+! Now call double precision fesft
+  CALL fesft_dp                                                               &
+      (pti_dp,      pdp_dp,      pclc_dp,     pwv_dp,      psw_dp,            &
+       pqlwc_dp,    pqiwc_dp,    pduco2_dp,   pduo3_dp,                       &
+       paeq1_dp,    paeq2_dp,    paeq3_dp,    paeq4_dp,    paeq5_dp,          &
+       papre_dp,    psmu0_dp,    palso_dp,    palth_dp,    pskyview_dp,       &
+       pfcor_dp,    psig_dp,     psct_dp,                                     &
+       ki1sd,       ki1ed,                                 ki3sd,      ki3ed, &
+       ki1sc,       ki1ec,                                 ki3sc,      ki3ec, &
+       lsolar,      lcrf,        lradtopo,    idebug,      jindex,            &
+       pflt_dp,     pfls_dp,     pflt_s_dp,   pfls_s_dp,   pflsdir_dp,        &
+       pfltd_dp,    pfltu_dp,    pflsd_dp,    pflsu_dp,    pflsp_dp,          &
+       pflpar_dp,   pflsu_par_dp, pflsd_par_dp, pflsp_par_dp)
+#endif
 
 ! Transform output variables back to single precision
   pflt     (ki1sc:ki1ec,ki3sc:ki3ec+1)= REAL(pflt_dp     (ki1sc:ki1ec,ki3sc:ki3ec+1), sp)
@@ -4478,7 +4624,20 @@
 !==============================================================================
 !+ Module procedure in "Radiation" 
 !------------------------------------------------------------------------------
-
+#ifdef COUP_OAS_COS
+!CPS we receive albedo for direct radiation also (palp)
+SUBROUTINE fesft_dp(                                                     &
+       pti   , pdp  , pclc  , pwv  , psw  , pqlwc, pqiwc, pduco2, pduo3, &
+       paeq1 , paeq2, paeq3 , paeq4, paeq5,                              &
+       papre , psmu0, palso , palth, pskyview, pfcor,                    &
+       psig  , psct , palp  ,                                            &
+       ki1sd , ki1ed,                ki3sd, ki3ed,                       &
+       ki1sc , ki1ec,                ki3sc, ki3ec,                       &
+       lsolar, lcrf , lradtopo, idebug, jindex,                          &
+       pflt  , pfls , pflt_s, pfls_s, pflsdir,                           &
+       pfltd , pfltu, pflsd , pflsu, pflsp,                              &
+       pflpar, pflsu_par, pflsd_par, pflsp_par)
+#else
 SUBROUTINE fesft_dp(                                                     &
        pti   , pdp  , pclc  , pwv  , psw  , pqlwc, pqiwc, pduco2, pduo3, &
        paeq1 , paeq2, paeq3 , paeq4, paeq5,                              &
@@ -4490,8 +4649,7 @@
        pflt  , pfls , pflt_s, pfls_s, pflsdir,                           &
        pfltd , pfltu, pflsd , pflsu, pflsp,                              &
        pflpar, pflsu_par, pflsd_par, pflsp_par)
-
-
+#endif
 !------------------------------------------------------------------------------
 !
 ! Description:
@@ -4579,7 +4737,10 @@
     psmu0  (ki1sd:ki1ed)                              , & ! (1)
     palth  (ki1sd:ki1ed)                              , & ! (1)
     palso  (ki1sd:ki1ed)                              , & ! (1)
-  
+#ifdef COUP_OAS_COS
+    palp   (ki1sd:ki1ed)                              , & ! (1)
+         ! Solar surface albedo for parallel radiation
+#endif  
     ! External data and radiation corrections
     pskyview(ki1sd:ki1ed)                             , & !
     pfcor   (ki1sd:ki1ed)                
@@ -4686,7 +4847,9 @@
   REAL    (KIND=dp)        ::  &
   pbbr     (ki1sd:ki1ed,            ki3sd:ki3ed+1), & ! (W/m**2) Black body radiation at layer boundaries
   pflpt    (ki1sd:ki1ed)                          , & ! Solar flux at TOA
+#if ! defined COUP_OAS_COS
   palp     (ki1sd:ki1ed)                          , & ! Solar surface albedo for parallel radiation
+#endif
   pqsmu0   (ki1sd:ki1ed)                          , & ! Inverse of cosine of zenith angle
   palogt   (ki1sd:ki1ed,            ki3sd:ki3ed)  , & ! ln T
   palogp   (ki1sd:ki1ed,            ki3sd:ki3ed)  , & ! ln p
@@ -5313,9 +5476,11 @@
 ! DO j2 = ki2sc, ki2ec
     DO j1 = ki1sc, ki1ec
       pqsmu0(j1   ) = 1.0_dp / psmu0(j1   )
+#if ! defined COUP_OAS_COS
       palp  (j1   ) = (1.0_dp +                                           &
         0.5_dp * (psmu0(j1   ) * (1.0_dp/palso(j1   ) - 1.0_dp))) &
      / (1.0_dp + (psmu0(j1   ) * (1.0_dp/palso(j1   ) - 1.0_dp)))**2
+#endif
     ENDDO
 ! ENDDO
 
