--- /p/home/jusers/ghasemi1/jureca/project/RTDO1/terrsysmp/parflow3_2/pfsimulator/parflow_lib/pf_pfmg.c	2016-07-08 19:25:18.000000000 +0200
+++ /p/home/jusers/ghasemi1/jureca/project/RTDO1/terrsysmp/bldsva/intf_oas3/parflow3_2/tsmp/pf_pfmg.c	2019-12-04 12:52:08.064341000 +0100
@@ -295,8 +295,8 @@
       }
 
       /* Set the HYPRE grid */
-      HYPRE_StructGridCreate(MPI_COMM_WORLD, 3, &(instance_xtra->hypre_grid) );
-
+//CPS fix
+      HYPRE_StructGridCreate(amps_CommWorld, 3, &(instance_xtra->hypre_grid) );
       /* Set local grid extents as global grid values */
       ForSubgridI(sg, GridSubgrids(grid))
       {
@@ -348,7 +348,7 @@
       symmetric = MatrixSymmetric(pf_Bmat);
       if ( !(instance_xtra->hypre_mat) )
       {
-         HYPRE_StructMatrixCreate(MPI_COMM_WORLD, instance_xtra->hypre_grid, 
+         HYPRE_StructMatrixCreate(amps_CommWorld, instance_xtra->hypre_grid, 
 			       instance_xtra->hypre_stencil,
 			       &(instance_xtra->hypre_mat) );
 	 HYPRE_StructMatrixSetNumGhost(instance_xtra->hypre_mat, full_ghosts);
@@ -359,7 +359,7 @@
       /* Set up new right-hand-side vector */
       if ( !(instance_xtra->hypre_b) )
       {
-         HYPRE_StructVectorCreate(MPI_COMM_WORLD, 
+         HYPRE_StructVectorCreate(amps_CommWorld, 
 			       instance_xtra->hypre_grid, 
 			       &(instance_xtra->hypre_b) );
 	 HYPRE_StructVectorSetNumGhost(instance_xtra->hypre_b, no_ghosts);
@@ -369,7 +369,7 @@
       /* Set up new solution vector */
       if ( !(instance_xtra->hypre_x) )
       {
-         HYPRE_StructVectorCreate(MPI_COMM_WORLD, 
+         HYPRE_StructVectorCreate(amps_CommWorld, 
 			       instance_xtra->hypre_grid, 
 			       &(instance_xtra->hypre_x) );
 	 HYPRE_StructVectorSetNumGhost(instance_xtra->hypre_x, full_ghosts);
@@ -639,9 +639,8 @@
       EndTiming(public_xtra->time_index_copy_hypre);
 
       /* Set up the PFMG preconditioner */
-      HYPRE_StructPFMGCreate(MPI_COMM_WORLD,
+      HYPRE_StructPFMGCreate(amps_CommWorld,
 				&(instance_xtra->hypre_pfmg_data) );
-
       HYPRE_StructPFMGSetTol(instance_xtra->hypre_pfmg_data, 1.0e-30);
       /* Set user parameters for PFMG */
       HYPRE_StructPFMGSetMaxIter(instance_xtra->hypre_pfmg_data, max_iter);
