--- program_off.F90	2007-05-01 18:22:47.000000000 +0200
+++ program_off.F90	2019-12-04 12:52:05.675512000 +0100
@@ -59,6 +59,10 @@
   use spmdMod  
   use ESMF_Mod
   use perf_mod
+
+#if (defined COUP_OAS_COS || defined COUP_OAS_PFL)
+  use oas_clm_vardef , only : kl_comm
+#endif
 !
 ! !ARGUMENTS:
     implicit none
@@ -80,20 +84,53 @@
   logical  :: mpi_running  ! true => MPI is initialized 
   integer  :: mpicom_glob  ! MPI communicator
 
-  character(len=SHR_KIND_CL) :: nlfilename = "lnd.stdin"
+  character(len=SHR_KIND_CL) :: nlfilename = "lnd.stdin" , istr
+!
+  integer :: ierror, lengths_of_types, i,counter,rank
+  logical :: flag
+  
 !-----------------------------------------------------------------------
 
   ! -----------------------------------------------------------------
   ! Initialize MPI
   ! -----------------------------------------------------------------
 
+#if (defined COUP_OAS_COS || defined COUP_OAS_PFL)
+  call oas_clm_init
+  call spmd_init(kl_comm)
+  call mct_world_init(1,kl_comm,mpicom,comp_id)
+#else
   call mpi_initialized (mpi_running, ier)
   if (.not. mpi_running) call mpi_init(ier)
-  mpicom_glob = MPI_COMM_WORLD
+
+  call MPI_COMM_RANK (MPI_COMM_WORLD, rank, ier)
+
+
+  i=0
+  counter=0
+  !FGa: enables multi instances. reads a non-negative instance number from instanceMap.txt and splits MPI_COMM_WORLD 
+  open (unit=4711, file='instanceMap.txt',access='sequential', form='formatted', status='old', action='read', iostat=ierror)
+  if(ierror==0) then
+    do while(counter<=rank)    
+      read(4711, '(A)') istr
+      if(counter .eq. rank) then
+        read(istr,'(I8)') i
+        call chdir("./tsmp_instance_"//trim(istr))
+        exit
+      else
+        counter=counter+1  
+      endif        
+    enddo
+    close(4711)
+  endif
+
+ 
+  call MPI_COMM_SPLIT(MPI_COMM_WORLD,i,i,mpicom_glob,ierror) 
   call spmd_init(mpicom_glob)
   call mct_world_init(1,mpicom_glob,mpicom,comp_id)
+#endif
 
-  call t_startf('init')
+!CPS  call t_startf('init')          !why
 
   ! -----------------------------------------------------------------
   ! Initialize ESMF (needed for time-manager)
@@ -109,6 +146,7 @@
   call t_initf(nlfilename, LogPrint=masterproc, Mpicom=mpicom, &
                MasterTask=masterproc)
 
+  call t_startf('init')           !CPS     
   ! -----------------------------------------------------------------
   ! Initialize Orbital parameters
   ! -----------------------------------------------------------------
@@ -169,10 +207,28 @@
      !  call t_barrierf('barrier1b',mpicom)
      ! Run
 
+#if defined COUP_OAS_PFL
+    !  received fields from ParFlow
+     call receive_fld_2pfl(nstep-1)
+    !
+#endif
+
      call clm_run1()
 
      !  call t_barrierf('barrier2b',mpicom)
 
+#if defined COUP_OAS_COS
+     ! send fields to Cosmo
+     call send_fld_2cos
+     !
+#endif
+
+#if defined COUP_OAS_PFL
+    !  send flux fields to ParFlow
+  call send_fld_2pfl
+    !
+#endif
+
      call clm_run2()
 
      !  call t_barrierf('barrierd2p',mpicom)
@@ -197,12 +253,18 @@
 
   if (masterproc) then
      write(6,*)'SUCCESFULLY TERMINATING CLM MODEL at nstep= ',get_nstep()
+     call flush(6)      !CPS
   endif
   call t_prf('timing_all',mpicom)
   call t_finalizef()
 
+#if (defined COUP_OAS_COS || defined COUP_OAS_PFL)
+  ! Let OASIS finalize run
+  call oas_clm_finalize
+#else
   ! Finalize ESMF
   call ESMF_Finalize()
+#endif
 
   stop
 end program program_off
