--- SoilTemperatureMod.F90	2007-04-26 20:50:19.000000000 +0200
+++ SoilTemperatureMod.F90	2019-12-04 12:52:05.691402000 +0100
@@ -795,6 +795,10 @@
                 heatr = 0._r8
                 if (xm(c,j) > 0._r8) then
                    h2osoi_ice(c,j) = max(0._r8, wice0(c,j)-xm(c,j))
+!CMS: to be consistent w/ pfl that doesn't account for soil freezing
+!#if defined COUP_OAS_PFL
+!                   h2osoi_ice(c,j) = 0._r8
+!#endif
                    heatr = hm(c,j) - hfus*(wice0(c,j)-h2osoi_ice(c,j))/dtime
                 else if (xm(c,j) < 0._r8) then
                    if (j <= 0) then
@@ -804,6 +808,10 @@
                          h2osoi_ice(c,j) = 0._r8
                       else
                          h2osoi_ice(c,j) = min(wmass0(c,j) - supercool(c,j),wice0(c,j)-xm(c,j))
+!CMS:to be consistent w/ pfl that doesn't account for soil freezing
+!#if defined COUP_OAS_PFL
+!                         h2osoi_ice(c,j) = 0._r8
+!#endif
                       endif
                    endif
                    heatr = hm(c,j) - hfus*(wice0(c,j)-h2osoi_ice(c,j))/dtime
