--- clmtypeInitMod.F90	2007-04-24 23:49:05.000000000 +0200
+++ clmtypeInitMod.F90	2019-12-04 12:52:05.734715000 +0100
@@ -857,6 +857,10 @@
     allocate(pps%u10(beg:end))
     allocate(pps%fv(beg:end))
     allocate(pps%ram1(beg:end))
+#ifdef COUP_OAS_COS
+    allocate(pps%rah1(beg:end)) !CPS
+    allocate(pps%raw1(beg:end)) !CPS
+#endif
     allocate(pps%slasun(beg:end))
     allocate(pps%slasha(beg:end))
     allocate(pps%lncsun(beg:end))
@@ -960,6 +964,10 @@
     pps%u10(beg:end) = nan
     pps%fv(beg:end) = nan
     pps%ram1(beg:end) = nan
+#ifdef COUP_OAS_COS
+    pps%rah1(beg:end) = nan !CPS
+    pps%raw1(beg:end) = nan !CPS
+#endif
     pps%slasun(beg:end) = nan
     pps%slasha(beg:end) = nan
     pps%lncsun(beg:end) = nan
@@ -1276,7 +1284,10 @@
     allocate(pes%t_ref2m_max_inst(beg:end))
     allocate(pes%q_ref2m(beg:end))
     allocate(pes%t_veg(beg:end))
-
+#ifdef COUP_OAS_COS
+    allocate(pes%t_sf(beg:end))   !CPS
+    allocate(pes%q_sf(beg:end))   !CPS
+#endif
     pes%t_ref2m(beg:end) = nan
     pes%t_ref2m_min(beg:end) = nan
     pes%t_ref2m_max(beg:end) = nan
@@ -1284,7 +1295,10 @@
     pes%t_ref2m_max_inst(beg:end) = nan
     pes%q_ref2m(beg:end) = nan
     pes%t_veg(beg:end) = nan
-
+#ifdef COUP_OAS_COS
+    pes%t_sf(beg:end) = nan      !CPS
+    pes%q_sf(beg:end) = nan      !CPS
+#endif
   end subroutine init_pft_estate_type
 
 !------------------------------------------------------------------------
@@ -1648,6 +1662,9 @@
     allocate(pwf%qflx_snowcap(beg:end))
     allocate(pwf%qflx_evap_veg(beg:end))
     allocate(pwf%qflx_tran_veg(beg:end))
+#if (defined COUP_OAS_PFL || defined COUP_OAS_COS)
+    allocate(pwf%pfl_flx_tran_veg(beg:end,1:nlevsoi)) !MS
+#endif
     allocate(pwf%qflx_evap_can(beg:end))
     allocate(pwf%qflx_evap_soi(beg:end))
     allocate(pwf%qflx_evap_tot(beg:end))
@@ -1663,6 +1680,9 @@
     pwf%qflx_snowcap(beg:end) = nan
     pwf%qflx_evap_veg(beg:end) = nan
     pwf%qflx_tran_veg(beg:end) = nan
+#if (defined COUP_OAS_PFL || defined COUP_OAS_COS)
+    pwf%pfl_flx_tran_veg(beg:end,1:nlevsoi) = nan !MS
+#endif
     pwf%qflx_evap_can(beg:end) = nan
     pwf%qflx_evap_soi(beg:end) = nan
     pwf%qflx_evap_tot(beg:end) = nan
@@ -1698,6 +1718,9 @@
     allocate(pcf%psnsun(beg:end))
     allocate(pcf%psnsha(beg:end))
     allocate(pcf%fpsn(beg:end))
+!MU (25.02.13)
+    allocate(pcf%fplres(beg:end))
+!MU (25.02.13)
     allocate(pcf%frm(beg:end))
     allocate(pcf%frmf(beg:end))
     allocate(pcf%frms(beg:end))
@@ -1831,6 +1854,9 @@
     pcf%psnsun(beg:end) = nan
     pcf%psnsha(beg:end) = nan
     pcf%fpsn(beg:end) = nan
+!MU (25.02.13)
+    pcf%fplres(beg:end) = nan
+!MU (25.02.13)
     pcf%frm(beg:end) = nan
     pcf%frmf(beg:end) = nan
     pcf%frms(beg:end) = nan
@@ -2258,6 +2284,9 @@
     allocate(cps%tkdry(beg:end,nlevsoi))
     allocate(cps%tksatu(beg:end,nlevsoi))
     allocate(cps%smpmin(beg:end))
+#if (defined COUP_OAS_PFL || defined COUP_OAS_COS)
+    allocate(cps%pfl_psi(beg:end,nlevsoi))
+#endif
     allocate(cps%hkdepth(beg:end))
     allocate(cps%wtfact(beg:end))
     allocate(cps%fracice(beg:end,nlevsoi))
@@ -2314,6 +2343,9 @@
     cps%tkdry(beg:end,1:nlevsoi) = nan
     cps%tksatu(beg:end,1:nlevsoi) = nan
     cps%smpmin(beg:end) = nan
+#if (defined COUP_OAS_PFL || defined COUP_OAS_COS)
+    cps%pfl_psi(beg:end,1:nlevsoi) = nan
+#endif
     cps%hkdepth(beg:end) = nan
     cps%wtfact(beg:end) = nan
     cps%fracice(beg:end,1:nlevsoi) = nan
@@ -2427,6 +2459,9 @@
     allocate(cws%h2osoi_liq(beg:end,-nlevsno+1:nlevsoi))
     allocate(cws%h2osoi_ice(beg:end,-nlevsno+1:nlevsoi))
     allocate(cws%h2osoi_vol(beg:end,1:nlevsoi))
+#if (defined COUP_OAS_PFL || defined COUP_OAS_COS)
+    allocate(cws%pfl_h2osoi_vol(beg:end,1:nlevsoi)) !MS
+#endif
     allocate(cws%h2osno_old(beg:end))
     allocate(cws%qg(beg:end))
     allocate(cws%dqgdT(beg:end))
@@ -2443,6 +2478,9 @@
     cws%h2osoi_liq(beg:end,-nlevsno+1:nlevsoi)= nan
     cws%h2osoi_ice(beg:end,-nlevsno+1:nlevsoi) = nan
     cws%h2osoi_vol(beg:end,1:nlevsoi) = nan
+#if (defined COUP_OAS_PFL || defined COUP_OAS_COS)
+    cws%pfl_h2osoi_vol(beg:end,1:nlevsoi) = nan !MS
+#endif
     cws%h2osno_old(beg:end) = nan
     cws%qg(beg:end) = nan
     cws%dqgdT(beg:end) = nan
@@ -2625,6 +2663,9 @@
     allocate(cwf%qflx_qrgwl(beg:end))
     allocate(cwf%qmelt(beg:end))
     allocate(cwf%h2ocan_loss(beg:end))
+#if (defined COUP_OAS_PFL || defined COUP_OAS_COS)
+    allocate(cwf%pfl_flx_total(beg:end,1:nlevsoi)) !MS
+#endif
 
     cwf%qflx_infl(beg:end) = nan
     cwf%qflx_surf(beg:end) = nan
@@ -2634,6 +2675,9 @@
     cwf%qflx_qrgwl(beg:end) = nan
     cwf%qmelt(beg:end) = nan
     cwf%h2ocan_loss(beg:end) = nan
+#if (defined COUP_OAS_PFL || defined COUP_OAS_COS)
+    cwf%pfl_flx_total(beg:end,1:nlevsoi) = nan !MS
+#endif
 
   end subroutine init_column_wflux_type
 
@@ -3073,10 +3117,16 @@
 !------------------------------------------------------------------------
     allocate(gwf%qchan2(beg:end))
     allocate(gwf%qchocn2(beg:end))
-
+#if (defined COUP_OAS_PFL || defined COUP_OAS_COS)
+    allocate(gwf%pfl_flx_total_gcell(beg:end,1:nlevsoi)) !MS
+#endif
+    
     gwf%qchan2(beg:end) = 0._r8
     gwf%qchocn2(beg:end) = 0._r8
-
+#if (defined COUP_OAS_PFL || defined COUP_OAS_COS)
+    gwf%pfl_flx_total_gcell(beg:end,1:nlevsoi) = 0._r8 !MS
+#endif
+  
   end subroutine init_gridcell_wflux_type
 
 end module clmtypeInitMod
