--- clmtype.F90	2007-04-26 20:50:19.000000000 +0200
+++ clmtype.F90	2019-12-04 12:52:05.683695000 +0100
@@ -144,6 +144,10 @@
    real(r8), pointer :: ftii(:,:)        	!down diffuse flux below veg per unit dif flx  (numrad)
    real(r8), pointer :: u10(:) 		        !10-m wind (m/s) (for dust model)
    real(r8), pointer :: ram1(:) 		!aerodynamical resistance (s/m)
+#ifdef COUP_OAS_COS
+   real(r8), pointer :: rah1(:)                 !aerodynamical resistance (s/m) !CPS heat
+   real(r8), pointer :: raw1(:)                 !aerodynamical resistance (s/m) !CPS moisture
+#endif
    real(r8), pointer :: fv(:) 		        !friction velocity (m/s) (for dust model)
    ! new variables for CN code
    real(r8), pointer :: slasun(:)     !specific leaf area for sunlit canopy, projected area basis (m^2/gC)
@@ -373,6 +377,10 @@
    real(r8), pointer :: t_ref2m_max_inst(:) !instantaneous daily max of average 2 m height surface air temp (K)
    real(r8), pointer :: q_ref2m(:)          !2 m height surface specific humidity (kg/kg)
    real(r8), pointer :: t_veg(:)            !vegetation temperature (Kelvin)
+#ifdef COUP_OAS_COS
+   real(r8), pointer :: t_sf(:)             !surface temperature (Kelvin) !CPS
+   real(r8), pointer :: q_sf(:)             !surface humidity (kg/kg) !CPS
+#endif
 end type pft_estate_type
 
 !----------------------------------------------------
@@ -579,6 +587,10 @@
    real(r8), pointer :: qflx_dew_grnd(:)  !ground surface dew formation (mm H2O /s) [+]
    real(r8), pointer :: qflx_sub_snow(:)  !sublimation rate from snow pack (mm H2O /s) [+]
    real(r8), pointer :: qflx_dew_snow(:)  !surface dew added to snow pack (mm H2O /s) [+]
+#if (defined COUP_OAS_PFL || defined COUP_OAS_COS)
+   real(r8), pointer :: pfl_flx_tran_veg(:,:) ! CMS: vegetation transpiration fluxes sent to ParFlow (mm H2O/s) 
+                                              !      at pft level for each soil layer
+#endif
 end type pft_wflux_type
 
 !----------------------------------------------------
@@ -588,6 +600,9 @@
    real(r8), pointer ::	psnsun(:) 	  !sunlit leaf photosynthesis (umol CO2 /m**2/ s)
    real(r8), pointer ::	psnsha(:) 	  !shaded leaf photosynthesis (umol CO2 /m**2/ s)
    real(r8), pointer ::	fpsn(:) 	  !photosynthesis (umol CO2 /m**2 /s)
+!MU (25.02.13)
+   real(r8), pointer :: fplres(:)         !plant respiration (umol CO2 /m**2 /s)
+!MU (25.02.13)
    ! variables in the following block are used by the 
    ! original CLM carbon cycle code, and are not used 
    ! in the CN code
@@ -884,6 +899,9 @@
    real(r8), pointer :: tkdry(:,:)            !thermal conductivity, dry soil (W/m/Kelvin) (nlevsoi) 
    real(r8), pointer :: tksatu(:,:)           !thermal conductivity, saturated soil [W/m-K] (new) (nlevsoi) 
    real(r8), pointer :: smpmin(:) 	      !restriction for min of soil potential (mm) (new)
+#if (defined COUP_OAS_PFL || defined COUP_OAS_COS)
+   real(r8), pointer :: pfl_psi(:,:)          !soil matrix potential received from ParFlow [mm]
+#endif
    real(r8), pointer :: gwc_thr(:) 	      !threshold soil moisture based on clay content
    real(r8), pointer :: mss_frc_cly_vld(:)    ![frc] Mass fraction clay limited to 0.20
    real(r8), pointer :: mbl_bsn_fct(:) 	      !??
@@ -952,6 +970,9 @@
    real(r8), pointer :: h2osoi_liq(:,:)       !liquid water (kg/m2) (new) (-nlevsno+1:nlevsoi)    
    real(r8), pointer :: h2osoi_ice(:,:)       !ice lens (kg/m2) (new) (-nlevsno+1:nlevsoi)    
    real(r8), pointer :: h2osoi_vol(:,:)       !volumetric soil water (0<=h2osoi_vol<=watsat) [m3/m3]  (nlevsoi)		  
+#if (defined COUP_OAS_PFL || defined COUP_OAS_COS)
+   real(r8), pointer :: pfl_h2osoi_vol(:,:)   !volumetric soil water calculated by pfl (0<=h2osoi_vol<=watsat) [m3/m3]  (nlevsoi) !CMS              
+#endif
    real(r8), pointer :: h2osno_old(:)         !snow mass for previous time step (kg/m2) (new)
    real(r8), pointer :: qg(:)                 !ground specific humidity [kg/kg]
    real(r8), pointer :: dqgdT(:)              !d(qg)/dT
@@ -1064,6 +1085,9 @@
    real(r8), pointer :: qflx_qrgwl(:) 	!qflx_surf at glaciers, wetlands, lakes
    real(r8), pointer :: qmelt(:) 	!snow melt [mm/s]
    real(r8), pointer :: h2ocan_loss(:) ! mass balance correction term for dynamic weights
+#if (defined COUP_OAS_PFL || defined COUP_OAS_COS)
+   real(r8), pointer :: pfl_flx_total(:,:) ! CMS: source/sink flux passed to ParFlow at the column level for each soil layer
+#endif
 end type column_wflux_type
 
 !----------------------------------------------------
@@ -1455,6 +1479,9 @@
    real(r8), pointer :: qchocn2(:)              !history file RTM river (channel) flow into ocean (m**3/s)
 !FTO
    type(column_wflux_type)::	cwf_a		!column-level water flux variables averaged to gridcell
+#if (defined COUP_OAS_PFL || defined COUP_OAS_COS)
+   real(r8), pointer :: pfl_flx_total_gcell(:,:) ! CMS: source/sink flux passed to ParFlow at the gridcell level for each soil layer
+#endif
 end type gridcell_wflux_type
 
 !----------------------------------------------------
