--- clm_varcon.F90	2006-11-17 05:44:27.000000000 +0100
+++ clm_varcon.F90	2019-12-04 12:52:05.693388000 +0100
@@ -79,12 +79,12 @@
   real(r8) :: cnfac  = 0.5_r8     !Crank Nicholson factor between 0 and 1
   real(r8) :: ssi    = 0.033_r8   !Irreducible water saturation of snow
   real(r8) :: wimp   = 0.05_r8    !Water impremeable if porosity less than wimp
-  real(r8) :: pondmx = 10.0_r8    !Ponding depth (mm)
+  real(r8) :: pondmx = 0.0_r8     !Ponding depth (mm)  !CPS was 10 mm earlier, causing inconsistent physics in upper soil layer
   ! 4/14/05: PET
   ! Adding isotope code
-  real(r8), parameter :: preind_atm_del13c = -6.0   ! preindustrial value for atmospheric del13C
-  real(r8), parameter :: preind_atm_ratio = SHR_CONST_PDB + (preind_atm_del13c * SHR_CONST_PDB)/1000.0  ! 13C/12C
-  real(r8) :: c13ratio = preind_atm_ratio/(1.0+preind_atm_ratio) ! 13C/(12+13)C preind atmosphere
+  real(r8), parameter :: preind_atm_del13c = -6.0_r8   ! preindustrial value for atmospheric del13C
+  real(r8), parameter :: preind_atm_ratio = SHR_CONST_PDB + (preind_atm_del13c * SHR_CONST_PDB)/1000.0_r8  ! 13C/12C
+  real(r8) :: c13ratio = preind_atm_ratio/(1.0_r8+preind_atm_ratio) ! 13C/(12+13)C preind atmosphere
   
   !------------------------------------------------------------------
   ! Initialize water type constants
