--- BareGroundFluxesMod.F90	2007-04-26 20:50:19.000000000 +0200
+++ BareGroundFluxesMod.F90	2019-12-04 12:52:05.672994000 +0100
@@ -129,7 +129,16 @@
     real(r8), pointer :: rssun(:)         ! sunlit stomatal resistance (s/m)
     real(r8), pointer :: rssha(:)         ! shaded stomatal resistance (s/m)
     real(r8), pointer :: ram1(:)          ! aerodynamical resistance (s/m)
+#ifdef COUP_OAS_COS
+    real(r8), pointer :: rah1(:)          ! aerodynamical resistance (s/m) !CPS heat
+    real(r8), pointer :: raw1(:)          ! aerodynamical resistance (s/m) !CPS moisture
+    real(r8), pointer :: t_sf(:)          ! surface temperature (Kelvin) !CPS 
+    real(r8), pointer :: q_sf(:)          ! surface humidity (kg/kg) !CPS  
+#endif
     real(r8), pointer :: fpsn(:)          ! photosynthesis (umol CO2 /m**2 /s)
+!MU (25.02.13)
+    real(r8), pointer :: fplres(:)        ! plant respiration (umol CO2 /m**2 /s)
+!MU (25.02.13)
     real(r8), pointer :: rootr(:,:)       ! effective fraction of roots in each soil layer
     real(r8), pointer :: rresis(:,:)      ! root resistance by layer (0-1)  (nlevsoi)	
 !
@@ -202,7 +211,6 @@
     thv => clm3%g%l%c%ces%thv
     beta => clm3%g%l%c%cps%beta
     zii => clm3%g%l%c%cps%zii
-    ram1 => clm3%g%l%c%p%pps%ram1
     cgrnds => clm3%g%l%c%p%pef%cgrnds
     cgrndl => clm3%g%l%c%p%pef%cgrndl
     cgrnd => clm3%g%l%c%p%pef%cgrnd
@@ -233,7 +241,17 @@
     psnsun => clm3%g%l%c%p%pcf%psnsun
     psnsha => clm3%g%l%c%p%pcf%psnsha
     fpsn => clm3%g%l%c%p%pcf%fpsn
+!MU (25.02.13)
+    fplres => clm3%g%l%c%p%pcf%fplres
+!MU (25.02.13)
 
+    ram1 => clm3%g%l%c%p%pps%ram1
+#ifdef COUP_OAS_COS
+    rah1 => clm3%g%l%c%p%pps%rah1  !CPS
+    raw1 => clm3%g%l%c%p%pps%raw1  !CPS
+    t_sf => clm3%g%l%c%p%pes%t_sf !CPS
+    q_sf => clm3%g%l%c%p%pes%q_sf !CPS
+#endif
     ! Filter pfts where frac_veg_nosno is zero
 
     fn = 0
@@ -349,12 +367,17 @@
        if (dqh(p) .gt. 0._r8) then   !dew
          rsoil = 0._r8
        else
-         rsoil   = (1._r8 - frac_sno(c)) * exp(8.206 - 4.255*www)
+         !CPSrsoil   = (1._r8 - frac_sno(c)) * exp(8.206 - 4.255*www)
+         rsoil   = (1._r8 - frac_sno(c)) * exp(8.206_r8 - 4.255_r8*www)
        end if
 
        raiw    = forc_rho(g)/(raw+rsoil)
-       ram1(p) = ram  !pass value to global variable
 
+       ram1(p) = ram          !pass value to global variable
+#ifdef COUP_OAS_COS
+       rah1(p) = rah          !CPS : pass value to global variable
+       raw1(p) = raw + rsoil  !CPS : pass value to global variable
+#endif
        ! Output to pft-level data structures
        ! Derivative of fluxes with respect to ground temperature
 
@@ -388,11 +411,19 @@
        rssun(p) = 1._r8/1.e15_r8 * cf
        rssha(p) = 1._r8/1.e15_r8 * cf
 
+#ifdef COUP_OAS_COS
+       ! Variablse needed for CLM-COSMO coupling
+       t_sf(p) = t_grnd(c)    !CPS
+       q_sf(p) = qg(c)        !CPS
+#endif
        ! Add the following to avoid NaN
 
        psnsun(p) = 0._r8
        psnsha(p) = 0._r8
        fpsn(p) = 0._r8
+!MU (25.02.13)
+       fplres(p) = 0._r8
+!MU (25.02.13)
        clm3%g%l%c%p%pps%lncsun(p) = 0._r8
        clm3%g%l%c%p%pps%lncsha(p) = 0._r8
        clm3%g%l%c%p%pps%vcmxsun(p) = 0._r8
